File: EA_TP1.py
from sys import stdin, stdout


class Pieces:
    numbers = []
    col = 0
    row = 0
    used = False

    def getPairsAndOccurrencies(self) -> None:
        for pi in range(len(self.numbers)):
            if b.occurencies.get(self.numbers[pi]) is None:
                b.occurencies[self.numbers[pi]] = 1
            else:
                b.occurencies[self.numbers[pi]] += 1

            if b.pairs.get((self.numbers[pi], self.numbers[(pi + 1) % 4])) is None:
                b.pairs[(self.numbers[pi], self.numbers[(pi + 1) % 4])] = [self]
            else:
                b.pairs[(self.numbers[pi], self.numbers[(pi + 1) % 4])].append(self)


class Board:
    n_pieces = rows = columns = 0
    pieces_arr = []
    board = []
    pairs = {}
    occurencies = {}

    def remove_piece(self, ind):
        self.board[ind].used = False
        self.board[ind].col = 0
        self.board[ind].row = 0
        self.board.pop(ind)

    def remove_pieces(self, ind):
        for i in range(len(self.board) - 1, ind - 1, -1):
            self.board[i].used = False
            self.board[i].row = 0
            self.board[i].col = 0
            self.board.pop(i)

    def printBoard(self):
        for r in range(self.rows):
            for piece in self.board:
                if piece.row == r + 1:
                    outln(piece.numbers[0])
                    outln(" ")
                    outln(piece.numbers[1])
                    if piece.col != self.columns:
                        outln("  ")
                    else:
                        outln("\n")

            for piece in self.board:
                if piece.row == r + 1:
                    outln(piece.numbers[3])
                    outln(" ")
                    outln(piece.numbers[2])
                    if piece.col != self.columns:
                        outln("  ")
                    else:
                        outln("\n")
            if r + 1 != self.rows:
                outln("\n")


def checkMatch(piece1: Pieces, col: int, row: int) -> int:
    if col == b.columns and row == b.rows:
        return 1

    if col < b.columns and row == 1:
        r = checkMatchRight(piece1, col, row)
        """if r == 0:
            return 0"""

    if row < b.rows and col == b.columns:
        r = checkMatchBellow(col, row)
        """if r == 0:
            return 0"""

    if 1 < row <= b.rows and 1 <= col <= b.columns:
        r = checkMatchRightBellow(piece1, col, row)
        """if r == 0:
            return 0"""

    return 0


def checkMatchRight(piece1: Pieces, col: int, row: int) -> int:
    above = piece1.numbers[1]
    bellow = piece1.numbers[2]

    if b.pairs.get((bellow, above)) is None:
        col -= 1
        b.remove_piece(b.board.index(piece1))
        return 0

    for match in b.pairs[(bellow, above)]:
        if match.used is False:
            match.numbers = match.numbers[match.numbers.index(above):] + \
                            match.numbers[:match.numbers.index(above)]

            col += 1
            match.col = col
            match.row = row
            match.used = True
            b.board.append(match)

            ret = checkMatch(match, col, row)
            if ret == 1:
                return 1
            """else:
                print("Continue right")
                col -= 1
                b.remove_piece(b.board.index(match))
                continue"""
    return 0


def checkMatchBellow(col: int, row: int) -> int:
    ind = (len(b.board) % b.columns) * row
    print("row", row, "col", col)
    print("indice", ind)
    print(len(b.board))
    print("-"*20)
    right = b.board[ind].numbers[2]
    left = b.board[ind].numbers[3]

    if b.pairs.get((left, right)) is None:
        row -= 1
        col = b.columns
        b.remove_pieces(ind)
        return 0
    for match in b.pairs[(left, right)]:
        if match.used is False:
            match.numbers = match.numbers[match.numbers.index(left):] + \
                            match.numbers[:match.numbers.index(left)]

            row += 1
            col = 1
            match.col = col
            match.row = row
            match.used = True
            b.board.append(match)

            ret = checkMatch(match, col, row)
            if ret == 1:
                return 1
            """else:
                print("Continue bellow")
                row -= 1
                col = b.columns
                b.remove_piece(b.board.index(match))
                continue"""
    return 0


def checkMatchRightBellow(piece1: Pieces, col: int, row: int) -> int:
    above = piece1.numbers[1]
    bellow = piece1.numbers[2]
    print(piece1.numbers)
    possible_vertical = b.pairs.get((bellow, above))

    if possible_vertical is None:
        col -= 1
        b.remove_piece(b.board.index(piece1))
        return 0

    ind = (len(b.board) % b.columns) * (row - 1)
    print("row", row-1, "col", col+1)
    print("indice", ind)
    print(len(b.board))
    print(b.board[ind].numbers)
    print()
    right = b.board[ind].numbers[2]
    left = b.board[ind].numbers[3]

    possible_horizontal = b.pairs.get((left, right))

    if possible_horizontal is None:
        col -= 1
        b.remove_piece(b.board.index(piece1))
        return 0

    intersection = list(set(possible_horizontal) & set(possible_vertical))
    if intersection is []:
        col -= 1
        b.remove_piece(b.board.index(piece1))
        return 0
    print(f"bellow {bellow} - above {above} - left {left} - right {right}")

    for match in intersection:
        if match.used is False:

            print("piece", p.numbers)
            print("*"*20)
            match.numbers = match.numbers[match.numbers.index(above):] + \
                            match.numbers[:match.numbers.index(above)]
            col += 1
            match.col = col
            match.row = row
            match.used = True
            b.board.append(match)

            ret = checkMatch(match, col, row)
            if ret == 1:
                return 1
            """else:
                print("Continue right bellow")
                col -= 1
                b.remove_piece(b.board.index(match))
                continue"""
    return 0


def reset():
    global p
    p = Pieces()
    b.remove_pieces(0)
    for i in range(b.n_pieces - 1, -1, -1):
        b.pieces_arr.pop(i)
    b.pairs = {}
    b.occurencies = {}
    b.rows = b.columns = b.n_pieces = 0


def readln():
    return stdin.readline().rstrip()


def outln(s1):
    stdout.write(str(s1))


if __name__ == "__main__":
    n = int(input())
    b = Board()

    for _ in range(n):
        b.n_pieces, b.rows, b.columns = map(int, input().split())

        for i in range(b.n_pieces):
            p = Pieces()
            p.numbers = list(map(int, input().split()))
            b.pieces_arr.append(p)

            p.getPairsAndOccurrencies()

            if i == 0:
                p.used = True
                p.row = 1
                p.col = 1
                b.board.append(p)

        odd = 0
        for o in b.occurencies:
            if b.occurencies[o] % 2 != 0:
                odd += 1
            if odd > 4:
                break
        if odd > 4:
            outln("impossible puzzle!\n")
            reset()
            continue

        r = checkMatch(b.board[0], 1, 1)

        if len(b.board) == b.n_pieces:
            b.printBoard()
        else:
            outln("impossible puzzle!\n")
        reset()

File: EA_TP1.cpp
#include <iostream>
#include <vector>
#include <array>
#include <map>
#include <algorithm>

using namespace std;

class Piece
{
public:
    vector<int> num;
    int row;
    int col;
    bool used;
    bool tested;

    Piece(int n0, int n1, int n2, int n3)
    {
        num.push_back(n0);
        num.push_back(n1);
        num.push_back(n2);
        num.push_back(n3);

        row = 0;
        col = 0;
        used = false;
        tested = false;
    }

    void Rotate90()
    {
        int aux = num[1];
        num[1] = num[0];
        num[0] = num[3];
        num[3] = num[2];
        num[2] = aux;
    }

    void Rotate180()
    {
        int aux = num[2];
        num[2] = num[0];
        num[0] = aux;
        aux = num[3];
        num[3] = num[1];
        num[1] = aux;
    }

    void Rotate270()
    {
        int aux = num[3];
        num[3] = num[0];
        num[0] = num[1];
        num[1] = num[2];
        num[2] = aux;
    }

    pair<int, int> getBellowAbove()
    {
        return {num[2], num[1]};
    }

    pair<int, int> getLeftRight()
    {
        return {num[3], num[2]};
    }
};

class Board
{
public:
    int n_pieces;
    int rows;
    int columns;
    vector<Piece *> pieces; // um vetor funciona como um Arraylist em Java, nao precisa de ter tamanho definido
    vector<Piece *> board;

    Board(int np, int r, int c)
    {
        n_pieces = np;
        rows = r;
        columns = c;
        pieces.resize(np); // define o tamanho do vetor de pecas
        // nao se faz o mesmo com o da board para ter metodo de comparacao
        // com o tamanho final do mesmo e o numero existente de pecas
    }

    void removeLeftPiece()
    {
        board.back()->used = false;
        board.back()->col = board.back()->row = 0;
        board.pop_back();
    }

    void removeToALimit(int index)
    {
        for (int i = board.size() - 1; i >= index; i--)
        {
            board[i]->used = false;
            board[i]->col = board[i]->row = 0;
            board.pop_back();
        }
    }

    void printBoard()
    {
        for (int row = 1; row <= this->rows; row++)
        {
            for (Piece *piece : this->board)
            {
                if (piece->row == row)
                {
                    cout << piece->num[0] << " " << piece->num[1];
                    if (piece->col != columns)
                        cout << "  ";
                    else
                        cout << endl;
                }
            }
            for (Piece *piece : this->board)
            {
                if (piece->row == row)
                {
                    cout << piece->num[3] << " " << piece->num[2];
                    if (piece->col != columns)
                        cout << "  ";
                    else
                        cout << endl;
                }
            }
            if (row != rows)
                cout << endl;
        }
    }
};

map<pair<int, int>, vector<Piece *>> pairs; // equivalente a dicionarios em python
map<int, int> occurrences;

void getPairsAndOccurrencies(Piece *p);

bool tree(Board *b, int row, int col);

bool addPrimeiraLinha(Board *b, int row, int col, pair<int, int> p);

bool addPrimeiraColuna(Board *b, int row, int col, pair<int, int> p);

bool addPeca(Board *b, int row, int col, vector<Piece *> intersection, vector<int> trio);

vector<Piece *> Intersection(vector<Piece *> &nums1, vector<int> trio);

void reset(Board *b);

void getPairsAndOccurrencies(Piece *p)
{
    // percorre a peca e conta as occorencias de cada numero
    // cria ainda os 4 pares possiveis na peca
    for (int i = 0; i < 4; i++)
    {
        if (occurrences.find(p->num[i]) == occurrences.end())
            occurrences.insert({p->num[i], 1});
        else
            occurrences[p->num[i]]++;

        // o pair cria uma lista com 2 elementos que vai funcionar como key do map
        pair<int, int> pp = {p->num[i], p->num[(i + 1) % 4]};
        if (pairs.find(pp) == pairs.end())
            pairs[pp] = {p};
        else
            pairs[pp].push_back(p);
    }
}

bool tree(Board *b, int row, int col)
{
    Piece *piece = NULL;
    Piece *piece2 = NULL;

    if (b == NULL || b->board.empty())
        return false; // verifica se o ponteiro para a Board existe ou nao e verifica se o vetor de pecas esta vazio

    // peca esta na primeira linha
    if (row == 1 && col <= b->columns)
    {
        piece = b->board[col - 2];

        // nao foi encontrada nenhuma peca com a coluna e a linha pretendida
        if (piece == NULL)
            return false;

        pair<int, int> p = piece->getBellowAbove();
        // se o par da peca da esquerda nao existir
        if (pairs.find(p) == pairs.end())
            return false;

        if (addPrimeiraLinha(b, 1, col, p))
            return true;
    }
    else if (row > 1 && row <= b->rows && col == 1)
    {
        // get peca de cima
        piece2 = b->board[(row - 2) * (b->columns)];

        // nao foi encontrada nenhuma peca de cima para as linhas depois da primeira
        if (piece2 == NULL)
            return false;

        pair<int, int> p2 = piece2->getLeftRight();

        //  se o par da peca de cima nao existir
        if (pairs.find(p2) == pairs.end())
            return false;

        if (addPrimeiraColuna(b, row, 1, p2)) // nao esta a aplicar a rotacao correta
            return true;

    }
    else if (row > 1 && row <= b->rows && col > 1 && col <= b->columns)
    {
        // get peca de cima e da esquerda
        piece = b->board[(row - 1) * b->columns + col - 2];      // peca da esquerda
        piece2 = b->board[(row - 2) * (b->columns) + (col - 1)]; // peca de cima

        // nao foi encontrada nenhuma peca com a coluna e a linha pretendida
        // nao foi encontrada nenhuma peca de cima para as linhas depois da primeira
        if (piece == NULL || piece2 == NULL)
            return false;

        pair<int, int> p = piece->getBellowAbove();
        // se o par da peca da esquerda nao existir
        if (pairs.find(p) == pairs.end())
            return false;

        pair<int, int> p2 = piece2->getLeftRight();
        // se o par da peca de cima nao existir
        if (pairs.find(p2) == pairs.end())
            return false;

        // vector com pecas que encaixam segundo o par da peca da esquerda
        vector<Piece *> possibleLeft = pairs[p];

        // {bellow, above=left, right}
        vector<int> trio = {p.first, p.second, p2.second};

        // intersecao dos 2 vectors
        vector<Piece *> intersection = Intersection(possibleLeft, trio);

        if (intersection.empty())
            return false;

        /*for (auto &i : intersection)
        {
            cout << i->num[0] << " " << i->num[1] << " " << i->num[2] << " " << i->num[3] << endl;
        }
        cout << endl;*/

        if (addPeca(b, row, col, intersection, trio)) // nao esta a aplicar a rotacao correta
            return true;

    }
    return false;
}

bool addPrimeiraLinha(Board *b, int row, int col, pair<int, int> p)
{
    for (auto &match : pairs[p])
    {
        if (!match->used)
        {

            if (match->num[3] == p.second && match->num[2] == p.first)
                match->Rotate90();

            if (match->num[2] == p.second && match->num[1] == p.first)
                match->Rotate180();

            if (match->num[1] == p.second && match->num[0] == p.first)
                match->Rotate270();

            match->col = col;
            match->row = row;
            match->used = true;

            b->board.push_back(match);

            if ((row * col) == b->n_pieces)
                return true;

            if (col == b->columns) // se estiver no fim da linha passa para a de baixo
            {
                row++;
                col = 1;
            }
            else
                col++;

            if (tree(b, row, col))
                return true;

            if (col == 1)
            { // se estiver no inicio da linha volta para o final da linha anterior
                row--;
                col = b->columns;
            }
            else
                col--;

            int pieceIndex = find(b->board.begin(), b->board.end(), match) - b->board.begin();
            b->removeToALimit(pieceIndex);
        }
    }

    return false;
}

bool addPrimeiraColuna(Board *b, int row, int col, pair<int, int> p)
{
    for (auto &match : pairs[p])
    {
        if (!match->used)
        {

            if (match->num[0] == p.second && match->num[3] == p.first)
                match->Rotate90();

            if (match->num[3] == p.second && match->num[2] == p.first)
                match->Rotate180();

            if (match->num[2] == p.second && match->num[1] == p.first)
                match->Rotate270();

            match->col = col;
            match->row = row;
            match->used = true;

            b->board.push_back(match);

            if ((row * col) == b->n_pieces)
                return true;

            if (b->columns == 1)
                row++;
            else
            {
                col++;
            }

            if (tree(b, row, col))
                return true;

            if (b->columns == 1)
                row--;
            else
                col--;

            int pieceIndex = find(b->board.begin(), b->board.end(), match) - b->board.begin();
            b->removeToALimit(pieceIndex);
        }
    }

    return false;
}

bool addPeca(Board *b, int row, int col, vector<Piece *> intersection, vector<int> trio)
{
    for (auto &match : intersection)
    {
        if (!match->used)
        {

            if (match->num[0] == trio[2] && match->num[3] == trio[1] && match->num[2] == trio[0])
                match->Rotate90();

            if (match->num[3] == trio[2] && match->num[2] == trio[1] && match->num[1] == trio[0])
                match->Rotate180();

            if (match->num[2] == trio[2] && match->num[1] == trio[1] && match->num[0] == trio[0])
                match->Rotate270();

            match->col = col;
            match->row = row;
            match->used = true;

            b->board.push_back(match);

            if ((row * col) == b->n_pieces)
                return true;

            // se estiver na ultima coluna passa para a linha de baixo
            if (col == b->columns)
            {
                col = 1;
                row++;
            }
            else
                col++;

            if (tree(b, row, col))
                return true;

            // se estiver na ultima coluna volta para a linha de cima
            if (col == 1)
            {
                row--;
                col = b->columns;
            }
            else
                col--;

            int pieceIndex = find(b->board.begin(), b->board.end(), match) - b->board.begin();
            b->removeToALimit(pieceIndex);
        }
    }
    return false;
}

vector<Piece *> Intersection(vector<Piece *> &nums1, vector<int> trio)
{

    sort(nums1.begin(), nums1.end()); // sort nums1 (non-decreasing order)

    vector<Piece *> res;

    for (auto &n : nums1)
    {
        for (int j = 0; j < 4; j++)
        {
            if (n->num[j] == trio[0] && n->num[(j + 1) % 4] == trio[1] && n->num[(j + 2) % 4] == trio[2])
            {
                res.push_back(n);
                break;
            }
        }
    }

    return res;
}

void reset(Board *b)
{
    b->board.clear();
    b->pieces.clear();
    pairs.clear();
    occurrences.clear();
    b->n_pieces = b->columns = b->rows = 0;
}

int main()
{
    ios_base::sync_with_stdio(0);
    cin.tie(0);

    int n = 0;

    cin >> n;

    for (int _ = 0; _ < n; _++)
    {
        int np = 0, r = 0, c = 0;

        cin >> np >> r >> c;
        Board *b = new Board(np, r, c);

        for (int i = 0; i < np; i++)
        {
            int n0, n1, n2, n3 = 0;
            cin >> n0 >> n1 >> n2 >> n3;
            Piece *p = new Piece(n0, n1, n2, n3);

            getPairsAndOccurrencies(p);

            if (i == 0)
            {
                p->col = p->row = 1;
                p->used = true;
                b->board.push_back(p);
            }

            b->pieces.push_back(p);
        }

        int odd = 0;
        for (auto &o : occurrences)
        {
            if (o.second % 2 != 0)
                odd++;
            if (odd > 4)
                break;
        }
        if (odd > 4)
        {
            cout << "impossible puzzle!\n";
            reset(b);
            continue;
        }

        bool res;
        if (b->columns == 1)
            res = tree(b, 2, 1);
        else
            res = tree(b, 1, 2);

        if (res)
            b->printBoard();
        else
            cout << "impossible puzzle!\n";
        reset(b);
    }
}